pkg/operator/controller/dns_status.go

```diff
@@ -92,34 +170,109 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 		switch {
 		case want == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsDesired")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsDesired, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.")
 		case have == 0:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsAvailable, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, "No DNS pods are available.")
 		case intstrErr != nil:
-			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
+			keepConditionUnknown = true
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSInvalidDNSMaxUnavailable, Status: operatorv1.ConditionUnknown, LastTransitionTime: now})
 			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
 		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
-			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSMaxUnavailableDNSPodsExceeded, Status: operatorv1.ConditionTrue, LastTransitionTime: now})
 			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
 		}
 	}
 
-	if len(degradedReasons) != 0 {
-		degradedCondition.Status = status
-		degradedCondition.Reason = strings.Join(degradedReasons, "")
-		degradedCondition.Message = strings.Join(messages, "\n")
-	} else {
-		degradedCondition.Status = operatorv1.ConditionFalse
-		degradedCondition.Reason = "AsExpected"
-		degradedCondition.Message = "Enough DNS pods are available, and the DNS service has a cluster IP address."
+	graceConditions, checkedConditions, requeueAfter := checkConditions(expectedConditions, degradedConditions)
+
+	// graceConditions have grace periods configured in the expectedConditions, so if present, this should be requeued
+	if len(graceConditions) != 0 {
+		finalDegradedCondition.Status = operatorv1.ConditionTrue
+		finalDegradedCondition.Reason = DNSControllerDegradedConditionType
+		finalDegradedCondition.Message = strings.Join(messages, "\n")
+		grace := formatConditions(graceConditions)
+
+		// If the oldCondition changed more recently than requeueAfter period of time, reduce the requeueAfter period of time
+		if oldCondition != nil && (!oldCondition.LastTransitionTime.IsZero()) {
+			t1 := now.Add(-requeueAfter)
+			t2 := oldCondition.LastTransitionTime
+			if t2.After(t1) {
+				requeueAfter = t2.Sub(t1)
+			}
+		}
```

Comment:
The purpose of this logic isn't clear to me.  `checkConditions` already checks each condition's last transition time...  But now I realize that the status conditions computed above all have "now" as their last transition time; is the logic here intended to adjust for that?  

I'm starting to think that it really would be simpler to put all the status conditions in the dns's actual status field so that we would have accurate last transition times.

Reply:
Me too - done